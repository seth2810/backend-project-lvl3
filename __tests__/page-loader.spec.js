import path from 'path';
import { promises as fs } from 'fs';

import nock from 'nock';

import downloadPage from '../index.js';

import * as helpers from './helpers/index.js';

describe('page-loader', () => {
  const pageFileName = 'ru-hexlet-io';
  const pageUrl = 'https://ru.hexlet.io';

  beforeAll(() => {
    nock.disableNetConnect();
  });

  beforeEach(async () => {
    nock.cleanAll();
  });

  it('should download page file', async () => {
    const outputDir = await helpers.makeOutputDir();

    nock(pageUrl).get('/').replyWithFile(200, helpers.resolveFixturePath('basic.html'));

    await downloadPage(pageUrl, outputDir);

    await expect(fs.readFile(path.join(outputDir, `${pageFileName}.html`), 'utf-8')).resolves.toMatchSnapshot();
  });

  it('should download page assets', async () => {
    const outputDir = await helpers.makeOutputDir();

    nock(pageUrl).get('/').replyWithFile(200, helpers.resolveFixturePath('assets.html'));

    nock(pageUrl).get('/assets/style.css').replyWithFile(200, helpers.resolveFixturePath('assets/style.css'));
    nock(pageUrl).get('/assets/image.jpg').replyWithFile(200, helpers.resolveFixturePath('assets/image.jpg'));
    nock(pageUrl).get('/assets/script.js').replyWithFile(200, helpers.resolveFixturePath('assets/script.js'));

    await downloadPage(pageUrl, outputDir);

    await expect(fs.readFile(path.join(outputDir, `${pageFileName}.html`), 'utf-8')).resolves.toMatchSnapshot();

    await expect(
      fs.readFile(path.join(outputDir, `${pageFileName}_files`, 'assets-style.css'), 'utf-8'),
    ).resolves.toBe(await fs.readFile(helpers.resolveFixturePath('assets/style.css'), 'utf-8'));

    await expect(
      fs.readFile(path.join(outputDir, `${pageFileName}_files`, 'assets-image.jpg'), 'utf-8'),
    ).resolves.toBe(await fs.readFile(helpers.resolveFixturePath('assets/image.jpg'), 'utf-8'));

    await expect(
      fs.readFile(path.join(outputDir, `${pageFileName}_files`, 'assets-script.js'), 'utf-8'),
    ).resolves.toBe(await fs.readFile(helpers.resolveFixturePath('assets/script.js'), 'utf-8'));
  });

  it('throws on network errors', async () => {
    const outputDir = await helpers.makeOutputDir();

    await expect(downloadPage(pageUrl, outputDir)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error when page unavailable', async () => {
    const outputDir = await helpers.makeOutputDir();

    nock(pageUrl).get('/').reply(404);

    await expect(downloadPage(pageUrl, outputDir)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error when output path not exists', async () => {
    const outputDir = '/notexitst';

    nock(pageUrl).get('/').replyWithFile(200, helpers.resolveFixturePath('basic.html'));

    await expect(downloadPage(pageUrl, outputDir)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error when asset not available', async () => {
    const outputDir = await helpers.makeOutputDir();

    nock(pageUrl).get('/').replyWithFile(200, helpers.resolveFixturePath('assets.html'));

    nock(pageUrl).get('/assets/style.css').reply(404);
    nock(pageUrl).get('/assets/image.jpg').replyWithFile(200, helpers.resolveFixturePath('assets/image.jpg'));
    nock(pageUrl).get('/assets/script.js').replyWithFile(200, helpers.resolveFixturePath('assets/script.js'));

    await expect(downloadPage(pageUrl, outputDir)).rejects.toThrow();
  });
});
