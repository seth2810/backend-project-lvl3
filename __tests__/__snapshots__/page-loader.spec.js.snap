// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page-loader should download page assets 1`] = `
"<!DOCTYPE html><html><head>
    <title>Example Domain</title>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <link href=\\"/courses\\" rel=\\"canonical\\">
    <link rel=\\"stylesheet\\" href=\\"ru-hexlet-io_files/assets-style.css\\">
    <link rel=\\"stylesheet\\" href=\\"https://cdn2.hexlet.io/packs/css/25-f5455698.chunk.css\\">
  </head>

  <body>
    <div>
      <h1>Example Domain</h1>
      <p>
        This domain is for use in illustrative examples in documents. You may
        use this domain in literature without prior coordination or asking for
        permission.
      </p>
      <img src=\\"ru-hexlet-io_files/assets-image.jpg\\">
      <p>
        <a href=\\"https://www.iana.org/domains/example\\">More information...</a>
      </p>
    </div>
    <script src=\\"ru-hexlet-io_files/assets-script.js\\"></script>
  

</body></html>"
`;

exports[`page-loader should download page file 1`] = `
"<!DOCTYPE html><html><head>
    <title>Example Domain</title>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <link href=\\"/courses\\" rel=\\"canonical\\">
  </head>

  <body>
    <div>
      <h1>Example Domain</h1>
      <p>
        This domain is for use in illustrative examples in documents. You may
        use this domain in literature without prior coordination or asking for
        permission.
      </p>
      <p>
        <a href=\\"https://www.iana.org/domains/example\\">More information...</a>
      </p>
    </div>
  

</body></html>"
`;

exports[`page-loader throws error when output path not exists 1`] = `"ENOENT: no such file or directory, open '/notexitst/ru-hexlet-io.html'"`;

exports[`page-loader throws error when page unavailable 1`] = `"Request failed with status code 404 (https://ru.hexlet.io)"`;

exports[`page-loader throws on network errors 1`] = `"Nock: Disallowed net connect for \\"ru.hexlet.io:443/\\" (https://ru.hexlet.io)"`;
